src/Hive.Versioning/Parsing/ParserErrorState.cs

```diff
@@ -0,0 +1,206 @@
+        /// <param name="left">The first report to compare.</param>
+        /// <param name="right">The second report to compare.</param>
+        /// <returns><see langword="true"/> if they are equal, <see langword="false"/> othwerwise.</returns>
+        public static bool operator ==(ActionErrorReport<TAction> left, ActionErrorReport<TAction> right)
+            => left.Equals(right);
+
+        /// <summary>
+        /// Compares two <see cref="ActionErrorReport{TAction}"/>s for inequality.
+        /// </summary>
+        /// <param name="left">The first report to compare.</param>
+        /// <param name="right">The second report to compare.</param>
+        /// <returns><see langword="true"/> if they are not equal, <see langword="false"/> othwerwise.</returns>
+        public static bool operator !=(ActionErrorReport<TAction> left, ActionErrorReport<TAction> right)
+            => !(left == right);
+
+        /// <inheritdoc/>
+        public override bool Equals(object? obj)
+            => obj is ActionErrorReport<TAction> report
+            && Equals(report);
+
+        /// <inheritdoc/>
+        public bool Equals(ActionErrorReport<TAction> other)
+            => EqualityComparer<TAction>.Default.Equals(Action, other.Action)
+            && TextOffset == other.TextOffset
+            && Length == other.Length;
+
+        /// <inheritdoc/>
+        public override int GetHashCode()
+            => HashCode.Combine(Action, TextOffset, Length);
+
+        /// <inheritdoc/>
+        public override string ToString() => $"({TextOffset},{TextOffset + Length}) {Action}";
+    }
+
+    /// <summary>
+    /// A type for recording parser errors in a memory-efficient manner.
+    /// </summary>
+    /// <remarks>
+    /// If this type is default constructed, errors will not be reported.
+    /// </remarks>
+    /// <typeparam name="TAction">The action type to use for reports.</typeparam>
+    public ref struct ParserErrorState<TAction>
+        where TAction : struct
+    {
+        /// <summary>
+        /// Gets whether or not this instance is reporting errors.
+        /// </summary>
+        public bool ReportErrors { get; internal set; }
+        /// <summary>
+        /// Gets the full input text associated with the parse this object is for.
+        /// </summary>
+        public StringPart InputText { get; internal set; }
+
+        /// <summary>
+        /// Constructs a new <see cref="ParserErrorState{TAction}"/> with the specified full text,
+        /// and enables error reporting for this instance.
+        /// </summary>
+        /// <param name="text"></param>
+        public ParserErrorState(in StringPart text)
+        {
+            ReportErrors = true;
+            InputText = text;
+            reports = default;
+            reports.Reserve(16);
+        }
+
+        private ArrayBuilder<ActionErrorReport<TAction>> reports;
+
+        private static long GetTextOffset(in StringPart start, in StringPart end)
```

Comment:
Is this unsafe block actually faster? I would have thought that the .Start call would be comparatively equivalent

Reply:
It is implemented like that in `!NETSTANDARD2_0` because in that scenario, `StringPart` is a `ReadOnlySpan<char>`, which *does not* have a way to get the offset from another string aside from comparing the pointers like this. Of course, this method assumes that the two parameters are in fact pointing into the same string, but that is something that both can't be easily verified, and is something that can be fairly easily guaranteed with a little bit of thought in the code.